### PR TITLE
LOOP-1680: Fix acknowledgement of repeating alerts 

### PR DESCRIPTION
**Pull request by rickpasetto**
Basically, we shouldn't be removing pending UNs upon acknowledgement.  That was a 1-line change, but I also did some refactoring that I'd meant to do a while ago.

https://tidepool.atlassian.net/browse/LOOP-1680

### CODE REVIEW

File: Loop/Managers/Alerts/AlertManager.swift [hunk 1]
```diff
@@ -89,10 +87,7 @@ extension AlertManager: AlertManagerResponder {
         if let responder = responders[identifier.managerIdentifier]?.value {
             responder.acknowledgeAlert(alertIdentifier: identifier.alertIdentifier)
         }
-        // Also clear the alert from the NotificationCenter
-        log.debug("Removing notification %@ from delivered & pending notifications", identifier.value)
-        userNotificationCenter.removeDeliveredNotifications(withIdentifiers: [identifier.value])
-        userNotificationCenter.removePendingNotificationRequests(withIdentifiers: [identifier.value])
```
Comment (rickpasetto) [inline Loop/Managers/Alerts/AlertManager.swift:95]: This line "removePendingNotificationRequests" was removed, but only after being moved below...

File: Loop/Managers/Alerts/UserNotificationAlertPresenter.swift [hunk 2]
```diff
@@ -44,3 +46,86 @@ class UserNotificationAlertPresenter: AlertPresenter {
         }
     }
 }
+
+extension UserNotificationAlertPresenter: AlertManagerResponder {
+    func acknowledgeAlert(identifier: Alert.Identifier) {
+        DispatchQueue.main.async {
+            self.log.debug("Removing notification %@ from delivered notifications", identifier.value)
+            self.userNotificationCenter.removeDeliveredNotifications(withIdentifiers: [identifier.value])
```
Comment (rickpasetto) [inline Loop/Managers/Alerts/UserNotificationAlertPresenter.swift:52]: ... to here.

**Review by nhamming — APPROVED** [review]
LGTM

File: Loop/Managers/Alerts/UserNotificationAlertPresenter.swift [hunk 3]
```diff
@@ -44,3 +46,86 @@ class UserNotificationAlertPresenter: AlertPresenter {
...
+        // TODO: Once we have a final design and approval for custom UserNotification buttons, we'll need to set categoryIdentifier
+//        userNotificationContent.categoryIdentifier = LoopNotificationCategory.alert.rawValue
+        userNotificationContent.threadIdentifier = identifier.value // Used to match categoryIdentifier, but I /think/ we want multiple threads for multiple alert types, no?
+        userNotificationContent.userInfo = [
+            LoopNotificationUserInfoKey.managerIDForAlert.rawValue: identifier.managerIdentifier,
+            LoopNotificationUserInfoKey.alertTypeID.rawValue: identifier.alertIdentifier,
+        ]
+        let encodedAlert = try encodeToString()
+        userNotificationContent.userInfo[AlertUserNotificationUserInfoKey.alert.rawValue] = encodedAlert
+        userNotificationContent.userInfo[AlertUserNotificationUserInfoKey.alertTimestamp.rawValue] =
```
Comment (ps2) [inline Loop/Managers/Alerts/UserNotificationAlertPresenter.swift]: I don't believe `userInfo[AlertUserNotificationUserInfoKey.alert.rawValue]` or `userInfo[AlertUserNotificationUserInfoKey.alert.rawValue]` are being used anywhere and can be removed, along with the alert string encoding methods.

**Review by ps2 — CHANGES_REQUESTED** [review]
In general looks good; I haven't tested, but QA should be able to test this. I think there is some code that should be deleted, though.

**Comment by ps2** [comment]
Is there anything using the encodeToString and decode(from string) methods anymore or can those be removed too?

**Comment by rickpasetto** [comment]
@ps2 Looks like they are not used, but they are also in LoopKit.  I don't feel strongly about removing them, since they seem like generally useful functions (and they are unit tested), but I'll put up a PR to remove if you feel strongly.

**Comment by ps2** [comment]
I usually have strong, happy feelings about removing code. :) I think it'd be a good idea, but isn't a huge priority.  It's code debt to leave it there.

**Review by ps2 — APPROVED** [review]
LGTM! Thanks!

**Comment by rickpasetto** [comment]
@ps2 Actually, while those functions are no longer used by this code, they are used by unit tests to test the `Codable` implementation which _is_ used elsewhere.  I think they're small and simple enough functions to be relatively harmless there in LoopKit, so I'd prefer to leave them...

**Comment by ps2** [comment]
Why would tests for Codable conformance use a custom string encoding?

Also, looks like `timestampFormatter` is no longer used either?

**Comment by rickpasetto** [comment]
@ps2 Custom? They just use JSON.  I could make it a bit cleaner by putting the extension into the tests instead... I'll do that.  See https://github.com/tidepool-org/LoopKit/pull/197

**Review by ps2 — APPROVED** [review]
I still don't understand why testing Codable conformance requires a separate encoding that isn't used in the main code, but maybe it's not important. Moving it into tests does help with the kruft aspect, as no-one will confuse it for the actual used encoding, so thanks!

**Pull request merged**